src/examples/libpmem2/unsafe_shutdowns.c

```diff
@@ -0,0 +1,429 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * unsafe_shutdowns.c -- unsafe shutdowns example for the libpmem2
+ */
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <libpmem2.h>
+
+#ifndef _WIN32
+#include <unistd.h>
+#else
+#include <io.h>
+#endif
+
+/* flags for struct shutdown_state */
+#define CLEAR_FLAG 0
+#define DIRTY_FLAG 1
+
+/*
+ * SDS_LEN -- overall struct shutdown_state len
+ */
+#define SDS_LEN 4096
+
+/*
+ * UUID_LEN -- length of the struct shutdown_state without fields: usc,
+ * flag, checksum (SDS_LEN - 8 - 1 - 8 = 4079)
+ */
+#define UUID_LEN 4079
+
+/*
+ * SDS_LEN_NO_CS -- length of the struct shutdown_state without checksum
+ * field
+ */
+#define SDS_LEN_NO_CS offsetof(struct shutdown_state, checksum)
+
+/*
+ * shutdown_state -- structure holding data about shutdown state
+ */
+struct shutdown_state {
+	uint64_t usc; /* unsafe shutdown count */
+	uint8_t uuid[UUID_LEN]; /* UID */
+	uint8_t flag;
+	uint64_t checksum;
+};
+
+/*
+ * checksum_compute -- calculate checksum of the structure using
+ * Fletcher's 64 algorithm
+ */
+static void
+checksum_compute(void *addr, size_t len, uint64_t *csump)
+{
+	if (len % 4 != 0)
+		abort();
+
+	uint32_t *p32 = addr;
+	uint32_t *p32end = (uint32_t *)((char *)addr + len);
+	uint32_t lo32 = 0;
+	uint32_t hi32 = 0;
+
+	while (p32 < p32end) {
+			lo32 = (lo32 + *p32) % 255;
+			hi32 = (hi32 + lo32) % 255;
+			++p32;
+	}
+
+	*csump = (uint64_t)hi32 << 32 | lo32;
+}
+
+/*
+ * prepare_environment -- prepare environment for create/read operation
+ */
+static void
+prepare_environment(int *fd, char *file, struct pmem2_config *cfg,
+		struct pmem2_map **map, void **mapping_addr,
+		size_t *mapping_size)
+{
+	if ((*fd = open(file, O_RDWR)) < 0) {
+		perror("open");
+		exit(1);
+	}
+
+	if (pmem2_config_new(&cfg)) {
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+
+	if (pmem2_config_set_fd(cfg, *fd)) {
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+
+	if (pmem2_config_set_required_store_granularity(cfg,
+			PMEM2_GRANULARITY_PAGE)) {
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+
+	if (pmem2_map(cfg, map)) {
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+
+	*mapping_addr = pmem2_map_get_address(*map);
+	*mapping_size = pmem2_map_get_size(*map);
+}
+
+/*
+ * cleanup -- cleaning up mapping, config and file descriptor
+ */
+static void
+cleanup(int *fd, struct pmem2_config *cfg, struct pmem2_map *map)
+{
+	pmem2_unmap(&map);
+	pmem2_config_delete(&cfg);
+	close(*fd);
+}
+
+/*
+ * create_data -- creating mapping and saving data to the mapping
+ */
+static void
+create_data(int *fd, char *file, struct pmem2_config *cfg,
```

Comment:
Try to minimize the difference between the "create" and "read" paths. As far as I can tell, there isn't any apart from the initial file creation, and the bunch of conditions in the read path (which will work fine on the create path also).